rust/src/conflict_checker.rs

```diff
@@ -0,0 +1,459 @@
+                Action::remove(remove) => Some(remove),
+                _ => None,
+            })
+            .collect()
+    }
+
+    pub fn added_files(&self) -> Vec<Add> {
+        self.actions
+            .iter()
+            .cloned()
+            .filter_map(|action| match action {
+                Action::add(add) => Some(add),
+                _ => None,
+            })
+            .collect()
+    }
+
+    pub fn blind_append_added_files(&self) -> Vec<Add> {
+        if self.is_blind_append().unwrap_or(false) {
+            self.added_files()
+        } else {
+            vec![]
+        }
+    }
+
+    pub fn changed_data_added_files(&self) -> Vec<Add> {
+        if self.is_blind_append().unwrap_or(false) {
+            vec![]
+        } else {
+            self.added_files()
+        }
+    }
+
+    pub fn only_add_files(&self) -> bool {
+        !self
+            .actions
+            .iter()
+            .any(|action| matches!(action, Action::remove(_)))
+    }
+
+    pub fn is_blind_append(&self) -> Option<bool> {
+        self.commit_info
+            .as_ref()
+            .map(|opt| opt.is_blind_append.unwrap_or(false))
+    }
+}
+
+pub(crate) struct ConflictChecker {
+    /// transaction information for current transaction at start of check
+    initial_current_transaction_info: CurrentTransactionInfo,
+    winning_commit_version: DeltaDataTypeVersion,
+    /// Summary of the transaction, that has been committed ahead of the current transaction
+    winning_commit_summary: WinningCommitSummary,
+    /// isolation level for the current transaction
+    isolation_level: IsolationLevel,
+    /// The state of the delta table at the base version from the current (not winning) commit
+    state: DeltaTableState,
+}
+
+impl ConflictChecker {
+    pub async fn try_new(
+        table: &DeltaTable,
+        winning_commit_version: DeltaDataTypeVersion,
+        operation: DeltaOperation,
+    ) -> Result<Self, DeltaTableError> {
+        // TODO raise proper error here
+        assert!(winning_commit_version == table.version() + 1);
+
+        // create winning commit summary
+        let commit_uri = table.commit_uri_from_version(winning_commit_version);
+        let commit_log_bytes = table.storage.get_obj(&commit_uri).await?;
+        let reader = BufReader::new(Cursor::new(commit_log_bytes));
+        let mut commit_actions = Vec::new();
+        for line in reader.lines() {
+            commit_actions.push(serde_json::from_str::<Action>(line?.as_str())?);
+        }
+        let winning_commit_summary =
+            WinningCommitSummary::new(commit_actions, winning_commit_version);
+
+        let initial_current_transaction_info = CurrentTransactionInfo::try_new(table, operation)?;
+
+        Ok(Self {
+            initial_current_transaction_info,
+            winning_commit_summary,
+            winning_commit_version,
+            isolation_level: IsolationLevel::Serializable,
```

Comment:
Why forcing the `IsolationLevel::Serializable`? What about the other isolation levels?